### ISSUE TITLE
leaking veths and sandboxes in 1.11.1

### ISSUE THREAD

**Issue by bprodoehl** [docker/docker#22397]
**Output of `docker version`:**

```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 9
 Running: 9
 Paused: 0
 Stopped: 0
Images: 16
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 2452
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.0-35-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.954 GiB
Name: docker-host.domain.com
ID: GX2B:QQCX:WX5I:KP7T:BIFJ:BDZH:KWJL:M6HO:3PUB:6MXJ:ZTZX:P3K6
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Virtual server

**Steps to reproduce the issue:**
1. Start and stop a lot of containers

**Describe the results you received:**
Eventually containers can't be started, because the bridge is full.  ifconfig | grep veth | wc -l reports that there are close to 1,000 veth devices, but only about 15 containers running.  Only a reboot will clean up the veths, and after a reboot, the docker log exhibits this behavior:

```
<lots of these "Removing stale sandbox" lines>
time="2016-04-29T00:21:45.883418406Z" level=info msg="Removing stale sandbox 8f9d8b4afc99e1ccf35774605b69d4a7f7d9aaf7c0d404e52194273de088cd96 (68667593e375190e37926d1773dc2f14c28a716db1f4a337d95f9a93c825844c)" 
time="2016-04-29T00:21:46.355371406Z" level=info msg="Removing stale sandbox 408f781df50dfaae11af8bb11bb946e75e3c682c49727b698df7c9e691d61316 (c89f6f77ca223c3763655ecf09c0b81bb90096d245b4d23420dd16f517945345)" 
<lots more of these "Removing stale endpoint" lines>
time="2016-04-29T00:23:08.521014816Z" level=info msg="Removing stale endpoint mycontainername (cc7d1d4e88e7f983cbad802c9a03b7db8721de6520b639c985bb28c34b585404)" 
time="2016-04-29T00:23:08.608674749Z" level=info msg="Removing stale endpoint anothercontainername (5dd8144891e748f5d60ec4a8a5fac4171de857815617375fecf46b9357990c77)" 
time="2016-04-29T00:23:08.704397341Z" level=info msg="Fixing inconsistent endpoint_cnt for network bridge. Expected=0, Actual=530" 
time="2016-04-29T00:23:19.952271389Z" level=fatal msg="Error starting daemon: Error initializing network controller: could not delete the default bridge network: network bridge has active endpoints" 
/var/run/docker.sock is up
time="2016-04-29T00:23:20.028766360Z" level=info msg="previous instance of containerd still alive (1085)" 
time="2016-04-29T00:23:20.118805751Z" level=info msg="[graphdriver] using prior storage driver \"aufs\"" 
time="2016-04-29T00:23:20.328700787Z" level=info msg="Graph migration to content-addressability took 0.00 seconds" 
time="2016-04-29T00:23:20.332567321Z" level=info msg="Firewalld running: false" 
time="2016-04-29T00:23:20.360464291Z" level=info msg="Fixing inconsistent endpoint_cnt for network bridge. Expected=0, Actual=18446744073709551593" 
time="2016-04-29T00:23:20.816060350Z" level=warning msg="Your kernel does not support swap memory limit." 
time="2016-04-29T00:23:20.816275541Z" level=warning msg="mountpoint for pids not found" 
time="2016-04-29T00:23:20.826830019Z" level=info msg="Loading containers: start." 
```

The tracking on the count in the bridge seems odd there, as well.  The first count, "Expected=0, Actual=530", seems believable for the leak, but then it fails, and gets "Expected=0, Actual=18446744073709551593".  I see that each time the system fails.

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by mavenugo** [comment]
@bprodoehl its not clear from the log if the daemon restarts again after the msg : 

`time="2016-04-29T00:23:19.952271389Z" level=fatal msg="Error starting daemon: Error initializing network controller: could not delete the default bridge network: network bridge has active endpoints" ` 

and the subsequent messages originates for the next time the daemon starts ? Also the insanely big number seems like a log formatting issue (hopefully :) ). But the daemon doesn't seem to have failed after this message is printed. Can you pls clarify ?

**Comment by bprodoehl** [comment]
Yes, it does appear that the daemon restarts in there, but from my perspective, it was all one run of starting the docker service, and I didn't snip any lines from between those.  It never sat and waited for me to try to start the docker service again.  I hope that makes sense.  After rebooting, the docker service does eventually get itself back to normal with a working bridge.  Maybe that insanely big number is a log formatting issue, but I'm just grabbing it out of the text file, and that message string is all nicely wrapped in quotes.

**Comment by crosbymichael** [comment]
I tried running 1024+ containers on my system and was not yet able to reproduce.   

@bprodoehl do you have any additional daemon settings other than the defaults?

**Comment by DamionWaltermeyer** [comment]
Hi @crosbymichael this is the relevant information from the machine @bprodoehl mentioned.

# Use DOCKER_OPTS to modify the daemon startup options.
DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4 --userland-proxy=false  --bip=172.17.42.1/16"

**Comment by crosbymichael** [comment]
@DamionWaltermeyer thanks, i'll try again with userland-proxy false.  There have been some kernel bugs with this and it could be the cause

**Comment by bprodoehl** [comment]
I think this may be caused by high load from the stats API endpoint.  If I stop all stats API requests, the daemon eventually reclaims the leaked veths over the course of a couple minutes.  I'm requesting container stats every 4 minutes, with 200ms between requests, but that still seems to be enough to essentially DoS the docker daemon.

**Comment by CVTJNII** [comment]
I believe I hit this intermittently on Docker 1.8.2, I would see veth devices left behind on machines with high container load.  As it was intermittent I was not able to determine root cause.  I was not doing anything direct API requests, just starting and removing containers (via docker run --rm) at a decent rate.  I've since upgraded to 1.10.3 and haven't seen it yet, though since this is filed against 1.11.1 that might be a coincidence.

**Comment by thomasem** [comment]
Seems related to https://github.com/docker/docker/issues/18814?

**Comment by bprodoehl** [comment]
I wasn't using the overlay network driver, so I don't think its related to #18814.

**Comment by thomasem** [comment]
@bprodoehl Gotcha. Okay, cool. I just remember having a similar issue there.

**Comment by ushuz** [comment]
Still seeing this with 18.06.1 on ubuntu 14.04.

**Comment by BenjaminBeck** [comment]
I have the same issue. 
Can i delete those /var/run/network/ifstateveth* files as a temporary, quick workaround?
